### DISCUSSION TITLE
Will the 130+ field limit be removed in the future and if so when?

### DISCUSSION THREAD

**Discussion by akselikap** [solspace/craft-freeform#205] [Questions & Answers] [answered]
Hi!
According to your website Freeform v4 would be coming out at the end of this year. According to [this](https://craftcms.stackexchange.com/a/32783) StackOverflow answer it should remove the limitation of possible fields but I couldn't really find more information regarding this. Also that answer states that Freeform v4 should've come out 1,5 years ago so that's why I'm asking about the schedule. Our use case requires more fields than the currently available amount so that's why this is important to us.

**Comment by kjmartens** [comment, marked as answer]
Hi @akselikap,

This is a great question. We pivoted from "going dark" for a long time with Freeform 4 development in favor of addressing a large number of issues incrementally in 3.9, 3.10, 3.11, 3.12 and likely a couple more. Freeform 4 won't be available this year - I will adjust the documentation again.

After putting out a survey, much to our surprise the vast majority of our customers that took the survey indicated they didn't have any issues with the current approach of shared fields across all forms and a large percentage preferred it that way. This meant that simply switching fields to be per form instead of global was not exactly going to be a slam dunk. Instead, we'll need to get a bit creative to allow both sides to be happy with the final solution. In the meantime, we did switch several more field types to use `TEXT` database type (vs VARCHAR), so it's far more likely that newer sites may get closer to 200-300 fields instead of around 100. Established sites can retroactively apply this by manually adjusting many Freeform fields in the `freeform_submissions` database table from `VARCHAR` to `TEXT`. 🙂

**Comment by ccchapman** [comment]
Thanks for the update.

Querying submissions seems to break once this issue is introduced i.e. once you exceed the field limit. Should Freeform handle this more gracefully e.g. catch the error when unable to alter the submissions table, drop the new field, and warn the user?

**Reply by kjmartens** [reply]
I last recall that there wasn't exactly an easy way to get it to behave this way (let the user know it failed), but Freeform 3.12 should resolve this from being an issue. :)

**Reply by ccchapman** [reply]
To clarify, the field limitation will be gone in Freeform 3.12?

**Reply by kjmartens** [reply]
As mentioned in my response above:

>...we did switch several more field types to use TEXT database type (vs VARCHAR), so it's far more likely that newer sites may get closer to 200-300 fields instead of around 100.

To clarify, we did not remove any field limitation (there will always be limitations out of our control), but it's far more likely that maxing out on fields won't be an issue for 99.9% of websites going forward from Freeform 3.12+ 🙂

**Reply by ccchapman** [reply]
Ok, it is still a limitation and issue in Freeform 3.12.

To my original point, I feel Freeform should catch the database exception when unable to alter the submissions table. Currently, you end up in a broken state where submissions cannot be loaded in the CP.

In my case, this broke the submissions in the CP _and_ the site itself due to a front-end submission query. It would be ideal if the content author couldn't break the site simply by adding too many Freeform fields.

**Reply by kjmartens** [reply]
Yes, we'll see if we can get an error to display when it happens as well :)
